src/Lean/Server/FileWorker/RequestHandling.lean

```diff
@@ -414,6 +414,87 @@ def handleSemanticTokensRange (p : SemanticTokensRangeParams)
   let endPos := text.lspPosToUtf8Pos p.range.end
   handleSemanticTokens beginPos endPos
 
+partial def handleFoldingRange (p : FoldingRangeParams)
+  : RequestM (RequestTask (Array FoldingRange)) := do
+  let doc ← readDoc
+  asTask do
+    let ⟨snaps, e?⟩ ← doc.allSnaps.updateFinishedPrefix
+    let mut stxs := snaps.finishedPrefix.map (·.stx)
+    match e? with
+    | some ElabTaskError.aborted =>
+      throw RequestError.fileChanged
+    | some (ElabTaskError.ioError e) =>
+      throw (e : RequestError)
+    | _ => pure ()
+
+    let lastSnap := snaps.finishedPrefix.getLastD doc.headerSnap
+    stxs := stxs ++ (← parseAhead doc.meta.text.source lastSnap).toList
+    let (_, ranges) ← StateT.run (addRanges doc.meta.text [] stxs) #[]
+    return ranges
+  where
+    isImport stx := stx.isOfKind ``Lean.Parser.Module.header || stx.isOfKind ``Lean.Parser.Command.open
+
+    addRanges (text : FileMap) sections
+    | [] => return
+    | stx::stxs => match stx with
+      | `(namespace $id)  => addRanges text (stx.getPos?::sections) stxs
+      | `(section $(id)?) => addRanges text (stx.getPos?::sections) stxs
+      | `(end $(id)?) => do
+        if let start::rest := sections then
+          addRange text FoldingRangeKind.region start stx.getTailPos?
+          addRanges text rest stxs
+        else
+          addRanges text sections stxs
+      | `(mutual $body* end) => do
+        addRangeFromSyntax text FoldingRangeKind.region stx
+        addRanges text [] body.toList
+        addRanges text sections stxs
+      | _ => do
+        if isImport stx then
+          addImportRanges text sections stx.getPos? stx.getTailPos? stxs
+        else
+          addCommandRange text stx
+          addRanges text sections stxs
+
+    addImportRanges (text : FileMap) sections startP endP stxs := do
+      if let stx::stxs := stxs then
+        if isImport stx then
+          return ← addImportRanges text sections (startP <|> stx.getPos?) (stx.getTailPos? <|> endP) stxs
+
+      addRange text FoldingRangeKind.imports startP endP
+      addRanges text sections stxs
```

Comment:
If we float this line into `addRanges`, every branch would end in `addRanges text _ stxs` and we could rewrite it as a `for` loop, correct? Just a suggestion, it's fine as is.

Reply:
The recursion here should also be avoidable using `stxs.partition isImport`

Reply:
> If we float this line into `addRanges`, every branch would end in `addRanges text _ stxs` and we could rewrite it as a `for` loop, correct? Just a suggestion, it's fine as is.

I don't think this is quite right, because sometimes our we recurse using `addImportRanges` and sometimes with `addRanges`, so it's unfortunately not quite a trivial for loop.

> The recursion here should also be avoidable using `stxs.partition isImport`

Do you mean `.span` rather than `.partition`? Yes, that might help a lot actually! - will give that a fiddle.

Reply:
Yes, looks good!